src/plugins/opencl_support/src/OpenCLSupportSettingsController.cpp

```diff
@@ -19,125 +19,155 @@
  * MA 02110-1301, USA.
  */
 
+#include "OpenCLSupportSettingsController.h"
+
 #include <QLabel>
 #include <QLayout>
-#include <QCheckBox>
 
-#include <U2Core/AppContext.h>
 #include <U2Algorithm/OpenCLGpuRegistry.h>
-#include <U2Core/AppResources.h>
 
-#include "OpenCLSupportSettingsController.h"
+#include <U2Core/AppContext.h>
+#include <U2Core/AppResources.h>
 
 namespace U2 {
 
-OpenCLSupportSettingsPageController::OpenCLSupportSettingsPageController( const QString & _displayMsg, QObject * p /* = 0 */ ) :
-AppSettingsGUIPageController( tr("OpenCL"), OpenCLSupportSettingsPageId, p ), displayMsg(_displayMsg) {}
+OpenCLSupportSettingsPageController::OpenCLSupportSettingsPageController(const QString &_displayMsg, QObject *p /* = 0 */)
+    : AppSettingsGUIPageController(tr("OpenCL"), OpenCLSupportSettingsPageId, p), displayMsg(_displayMsg) {
+}
 
-AppSettingsGUIPageState * OpenCLSupportSettingsPageController::getSavedState() {
+AppSettingsGUIPageState *OpenCLSupportSettingsPageController::getSavedState() {
     QList<OpenCLGpuModel *> registeredGpus = AppContext::getOpenCLGpuRegistry()->getRegisteredGpus();
-    OpenCLSupportSettingsPageState * s = new OpenCLSupportSettingsPageState( registeredGpus.size() );
-    for( int i = 0, end = s->enabledGpus.size(); i < end; ++i ) {
-        s->enabledGpus[i] = registeredGpus.at(i)->isEnabled();
+    QString nameOfEnabledGpu;
+    foreach (OpenCLGpuModel *gpu, registeredGpus) {
+        CHECK_CONTINUE(gpu->isEnabled());
+
+        nameOfEnabledGpu = gpu->getName();
+        break;
     }
 
-    return s;
+    return new OpenCLSupportSettingsPageState(nameOfEnabledGpu);
 }
 
-void OpenCLSupportSettingsPageController::saveState( AppSettingsGUIPageState * _s ) {
+void OpenCLSupportSettingsPageController::saveState(AppSettingsGUIPageState *_s) {
     QList<OpenCLGpuModel *> registeredGpus = AppContext::getOpenCLGpuRegistry()->getRegisteredGpus();
-    OpenCLSupportSettingsPageState * s = qobject_cast<OpenCLSupportSettingsPageState *>(_s);
+    OpenCLSupportSettingsPageState *s = qobject_cast<OpenCLSupportSettingsPageState *>(_s);
 
     //saving state of enabled/disabled GPUs into registry
-    for( int i = 0, end = s->enabledGpus.size(); i < end; ++i ) {
-        registeredGpus[i]->setEnabled( s->enabledGpus[i] );
+    const QString enabledGpu = s->getEnabledGpuName();
+    bool enabledGpuWasFound = false;
+    foreach (OpenCLGpuModel *gpu, registeredGpus) {
+        if (enabledGpu.startsWith(gpu->getName())) {
+            gpu->setEnabled(true);
+            enabledGpuWasFound = true;
+        } else {
+            gpu->setEnabled(false);
+        }
+    }
+    if (!enabledGpuWasFound) {
+        registeredGpus.first()->setEnabled(true);
     }
```

Comment:
Use method OpenCLGpuRegistry::getGpuByName.

Reply:
I'm not quite sure, that I can change a status of the enabled GPU only and everything will work well, it may cause an unexpected behaviour. It's better to set one GPU as enabled and the rest of GPUs as disabled.

Reply:
OK, as you wish.